static/js/message_fetch.js

```diff
@@ -315,14 +317,23 @@ exports.maybe_load_newer_messages = function (opts) {
     opts.show_loading();
     const anchor = exports.get_frontfill_anchor(msg_list).toFixed();
 
+    function load_more(args) {
+        if (args.fetch_again && args.msg_list === current_msg_list) {
+            exports.maybe_load_newer_messages({
+                msg_list: current_msg_list,
+                show_loading: message_scroll.show_loading_newer,
+                hide_loading: message_scroll.hide_loading_newer,
+            });
+        }
+    }
```

Comment:
I'm concerned that this `load_more` call will result in visiting a stream narrow fetching messages in a loop until we have the full history (at least downwards), even if you do a `near:1` narrow.  

Oh, I see, this is only going to happen with the `max_message_id` case.  Hmm.

OK, the other thing to worry about here is whether everything breaks with a horrible endless loop in the event that the `max_message_id` message gets deleted or edited to be in a different topic.

So I think we need a circuit-breaker here, which I think could be that just before doing this additional call, we clear the `max_message_id` state; that should be correct, since we expect to not see the race we're worried about here again.

(And you should add some comments explaining the purpose behind this feature so the codebase remains readable afterwards).

Reply:
Oh thanks for the review
Didn't think of this situation.
I'm resetting the `expected_max_message_id` back to 0 when the race condition occurs (Before the re-fetch).
But if a message is again received while this happens (the reason we are adding this re-fetch) the `expected_max_message_id`'s value again gets updated and we enter the fetch_status block again.

I'm looking into a solution where we have to re-fetch at most 1 time....